predict.py

```diff
@@ -0,0 +1,187 @@
+    "sen12ms": (SemanticSegmentationTask, SEN12MSDataModule),
+    "so2sat": (ClassificationTask, So2SatDataModule),
+    "ucmerced": (ClassificationTask, UCMercedDataModule),
+}
+
+
+def write_mask(mask: torch.Tensor, output_dir: str, input_filename: str) -> None:
+    """Write mask to specified output directory."""
+    output_path = os.path.join(output_dir, os.path.basename(input_filename))
+    with rio.open(input_filename) as src:
+        profile = src.profile
+    profile["count"] = 1
+    profile["dtype"] = "uint8"
+    mask = mask.cpu().numpy()
+    with rio.open(output_path, "w", **profile) as ds:
+        ds.write(mask)
+
+
+def main(config_dir: str, predict_on: str, output_dir: str, device: str) -> None:
+    """Main inference loop."""
+    os.makedirs(output_dir, exist_ok=True)
+
+    # Load checkpoint and config
+    conf = OmegaConf.load(os.path.join(config_dir, "experiment_config.yaml"))
+    ckpt = os.path.join(config_dir, "last.ckpt")
+
+    # Load model
+    task_name = conf.experiment.task
+    datamodule: pl.LightningDataModule
+    task: pl.LightningModule
+    if task_name not in TASK_TO_MODULES_MAPPING:
+        raise ValueError(
+            f"experiment.task={task_name} is not recognized as a valid task"
+        )
+    task_class, datamodule_class = TASK_TO_MODULES_MAPPING[task_name]
+    task = task_class.load_from_checkpoint(ckpt)
+    task = task.to(device)
+    task.eval()
+
+    # Load datamodule and dataloader
+    conf.experiment.datamodule["predict_on"] = predict_on
+    datamodule = datamodule_class(**conf.experiment.datamodule)
+    datamodule.setup()
+    dataloader = datamodule.predict_dataloader()
+
+    if len(os.listdir(output_dir)) > 0:
+        if conf.program.overwrite:
+            print(
+                f"WARNING! The output directory, {output_dir}, already exists, "
+                + "we will overwrite data in it!"
+            )
+        else:
+            raise FileExistsError(
+                f"The predictions directory, {output_dir}, already exists and isn't "
+                + "empty. We don't want to overwrite any existing results, exiting..."
+            )
+
+    for i, batch in enumerate(dataloader):
+        x = batch["image"].to(device)  # (N, B, C, H, W)
+        assert len(x.shape) in {4, 5}
+        if len(x.shape) == 5:
+            masks = []
+
+            def tensor_to_int(
+                tensor_tuple: Tuple[torch.Tensor, ...]
+            ) -> Tuple[int, ...]:
+                """Convert tuple of tensors to tuple of ints."""
+                return tuple(int(i.item()) for i in tensor_tuple)
+
+            original_shape = cast(
+                Tuple[int, int], tensor_to_int(batch["original_shape"])
+            )
+            patch_shape = cast(Tuple[int, int], tensor_to_int(batch["patch_shape"]))
+            padding = cast(Tuple[int, int], tensor_to_int(batch["padding"]))
+            patch_combine = CombineTensorPatches(
```

Comment:
Can you tell me more about how this function works? What happens if your patches have overlap? Is this sufficient to close https://github.com/microsoft/torchgeo/issues/30 or do we need something more powerful/generic?

Reply:
> Can you tell me more about how this function works? 

I've written a tutorial describing how extract and combine patches work [here](https://github.com/kornia/tutorials/blob/master/source/extract_combine_patches.ipynb)


> What happens if your patches have overlap? 

Currently you can **extract** patches with overlap (via the stride parameter) but you can't **merge** them together. This is because `CombineTensorPatches` currently only supports `stride=window_size` as seen [here](https://github.com/kornia/kornia/blob/ccd258d0956da89b1feca96448eff8e4969d405a/kornia/contrib/extract_patches.py#L247)

> Is this sufficient to close #30 or do we need something more powerful/generic?

Based on my understanding of the scope of #30, I would say no. If the goal is to just enable users to extract patches and stitch them together, kornia's `ExtractTensorPatches` and `CombineTensorPatches` are sufficient. Once `CombineTensorPatches` supports `stride!=window_size`, we will be able to handle patches with overlap.


But for the alternate stitching techniques (like label averaging) mentioned in the paper referenced in #30, we might need something more powerful as `CombineTensorPatches` doesn't support this.